### PR TITLE
Issue 154: BK Operator should support multiple ledger PVCs, multiple journal PVCs and multiple index PVCs for 1 bookie

### LINKED ISSUE #154: BK Operator should support multiple ledger PVCs, multiple journal PVCs and multiple index PVCs for 1 bookie
### Description

BK Operator should support multiple ledger PVCs, multiple journal PVCs and multiple index PVCs for 1 bookie

### Importance

should-have

### Location

pkg/controller/bookkeepercluster/bookie.go

### PR DESCRIPTION
**Pull request by CraneShiEMC**
### Change log description

BK Operator should support multiple ledger PVCs, multiple journal PVCs and multiple index PVCs for 1 bookie.
Now, if we set flag option **multiVolumesSetPerBookieEnabled** to true and we provide multiple ledger directories in `ledgerDirectories` options, for example, `ledgerDirectories: "/bk/ledgers/l0,/bk/ledgers/l1,/bk/ledgers/l2,/bk/ledgers/l3"`, for each ledger directory such as `/bk/ledgers/l0`, bk operator will create 1 PVC with VolumeMount mount to this ledger directory. So does `journalDirectories` and `indexDirectories` options.

By default, i.e. we don't set flag option **multiVolumesSetPerBookieEnabled** or we set this option to false, the bookkeeper operator will still have original behavior, i.e. 1 PVC with multiple volume mounts for multiple directories will be used.

### Purpose of the change

Fixes #154 

### What the code does

Now, if we set flag option **multiVolumesSetPerBookieEnabled** to true and provide multiple ledger directories in `ledgerDirectories` options, for example, `ledgerDirectories: "/bk/ledgers/l0,/bk/ledgers/l1,/bk/ledgers/l2,/bk/ledgers/l3"`, for each ledger directory such as `/bk/ledgers/l0`, bk operator will create 1 PVC with VolumeMount mount to this ledger directory. So does `journalDirectories` and `indexDirectories` options.

By default, i.e. we don't set flag option **multiVolumesSetPerBookieEnabled** or we set this option to false, the bookkeeper operator will still have original behavior, i.e. 1 PVC with multiple volume mounts for multiple directories will be used.

### How to verify it

Provide multiple directories for `ledgerDirectories`, `journalDirectories` or `indexDirectories` options and set flag option **multiVolumesSetPerBookieEnabled** to true while deploying bookies. Then, verify whether multiple PVCs for ledgers, journals or indices has been bound to each bookie as expected, one PVC mount to one directory. 

In our ObjectScale deployment with Pravega, bookie's multiple ledger PVCs, journal PVCs, index PVCs has been created and mount to multiple directories as expected. Also bookie has functioned normally and the IO can be balanced between multiple PVCs.

In addition, we have verified the negative test cases without **multiVolumesSetPerBookieEnabled** set and the bookie has original behavior as expected.

### CODE REVIEW

File: pkg/controller/bookkeepercluster/bookie.go [hunk 1]
```diff
@@ -365,29 +365,101 @@ func createVolumeMount(ledgerDirs []string, journalDirs []string, indexDirs []st
...
+	} else {
+		// default value if user did not set indexDirectories in options
+		indexDirs = append(indexDirs, "/bk/index")
+	}
+
+	var claims []corev1.PersistentVolumeClaim
+	if len(ledgerDirs) > 1 {
+		for i := 0; i < len(ledgerDirs); i++ {
+			name := LedgerDiskName + strconv.Itoa(i)
+			pvc := corev1.PersistentVolumeClaim{
```
Comment (anishakj) [inline pkg/controller/bookkeepercluster/bookie.go:396]: This is changing the current behavior of having one PVC with multiple mount points. We should make this option configurable.  for multiple PVCs  and we should make the default as one PVC with multiple mount points.

**Review by anishakj — CHANGES_REQUESTED** [review]

File: pkg/controller/bookkeepercluster/bookie.go (hunk 1 above)
Comment (CraneShiEMC) [reply pkg/controller/bookkeepercluster/bookie.go:396]: > This is changing the current behavior of having one PVC with multiple mount points. We should make this option configurable. for multiple PVCs and we should make the default as one PVC with multiple mount points.

OK. We will change the code to use a bk-op option, temporarily named as multiVolumesSetPerBookieEnabled, as a flag to control this behavior. By default or when it's set to false, the original one PVC with multiple mount points will be used. The multiple-PVC support will only be used when multiVolumesSetPerBookieEnabled is set to true.

File: pkg/controller/bookkeepercluster/bookie.go [hunk 2]
```diff
@@ -365,29 +401,106 @@ func createVolumeMount(ledgerDirs []string, journalDirs []string, indexDirs []st
 }
 
 func makeBookieVolumeClaimTemplates(bk *v1alpha1.BookkeeperCluster) []corev1.PersistentVolumeClaim {
-	return []corev1.PersistentVolumeClaim{
-		{
+	var ledgerDirs, journalDirs, indexDirs []string
+	var multiVolumesSetPerBookieEnabled bool
+	var ok bool
+
+	if _, ok = bk.Spec.Options["ledgerDirectories"]; ok {
```
Comment (anishakj) [inline pkg/controller/bookkeepercluster/bookie.go:408]: Lines 407-428 might not be required, if we are making multivolume support  to `false` . Could you please verify that?

**Review by anishakj — CHANGES_REQUESTED** [review]

**Review by anishakj — CHANGES_REQUESTED** [review]
Please add unit tests for the new code paths

File: pkg/controller/bookkeepercluster/bookie.go (hunk 2 above)
Comment (CraneShiEMC) [reply pkg/controller/bookkeepercluster/bookie.go:408]: > Lines 407-428 might not be required, if we are making multivolume support to `false` . Could you please verify that?

OK. I will optimize this part according to your suggestion.

**Review by SrishT — CHANGES_REQUESTED** [review]
@CraneShiEMC the changes look good.
I would however request you to increase code coverage and add documentation to explain how this option `multiVolumesSetPerBookieEnabled` should be configured to get the desired behaviour. Also please rebase your code (that would fix the end to end tests as well)
Could u also add a separate PR in the [charts](https://github.com/pravega/charts) repo, inside the bookkeeper charts, wherein the flag `multiVolumesSetPerBookieEnabled` has also been added with its default value (false) ?

**Comment by CraneShiEMC** [comment]
> @CraneShiEMC the changes look good.
> I would however request you to increase code coverage and add documentation to explain how this option `multiVolumesSetPerBookieEnabled` should be configured to get the desired behaviour. Also please rebase your code (that would fix the end to end tests as well)
> Could u also add a separate PR in the [charts](https://github.com/pravega/charts) repo, inside the bookkeeper charts, wherein the option `multiVolumesSetPerBookieEnabled` has also been added inside the bookkeeper options field with its default value (false) ?

OK! I will change the code according to your comments

**Comment by jkhalack** [comment]
Please, add a validation that `multiVolumesSetPerBookieEnabled` option is not changed on BK-cluster update. 
The best place to do it would be at the same time we are checking that ledger/journal directory paths have not changed: https://github.com/pravega/bookkeeper-operator/blob/master/pkg/apis/bookkeeper/v1alpha1/bookkeepercluster_types.go#L763-L779

**Pull request closed without merging**